### PR TITLE
(feat) Remote tokenizers

### PR DESCRIPTION
**Pull request by kbamponsem**
This PR adds remote tokenizers to BastionLab.

Currently, only HuggingFace Tokenizers are supported 

The PR adds the `RemoteTokenizer` and `tokenizers` module to the `bastionlab` python package.

`RemoteTokenizers` seeks to be very simple by forwarding/delegating important methods found in the `Tokenizer` class.

### CODE REVIEW

File: client/src/bastionlab/polars/remote_polars.py [hunk 1]
```diff
@@ -1327,3 +1396,84 @@ def __map(self: LDF, func, **kwargs) -> None:
 class RemoteLazyGroupBy(Generic[LDF]):
     _inner: pl.internals.lazyframe.groupby.LazyGroupBy[LDF]
     _meta: Metadata
+
+
+def train_test_split(
```
Comment (cchudant) [inline client/src/bastionlab/polars/remote_polars.py:1401]: is this supposed to be there or is this a mistake ? (#195)

File: client/src/bastionlab/torch/remote_torch.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    def nb_samples(self):
+        return self.labels.shape[0]
+
+    @staticmethod
+    def from_dataset(dataset: Dataset) -> "RemoteDataset":
+        data = dataset.__getitem__(0)
+        inputs = torch.cat(data[0]).unsqueeze(0)
+        labels = torch.tensor(data[1]).unsqueeze(0)
+
+        print("Dataset --> RemoteDataset Transformation")
```
Comment (cchudant) [inline client/src/bastionlab/torch/remote_torch.py:132]: use logging.trace or debug

File: client/src/bastionlab/torch/utils.py [hunk 3]
```diff
@@ -312,15 +313,14 @@ def serialize_dataset(
...
-        meta=bulk_serialize(
-            {
-                "input_shape": [input.size() for input in dataset[0][0]],
-                "input_dtype": [input.dtype for input in dataset[0][0]],
-                "nb_samples": len(dataset),  # type: ignore [arg-type]
-                "privacy_limit": privacy_limit,
-                "train_dataset": train_dataset,
-            }
-        ),
+        meta=bytes(),
```
Comment (cchudant) [inline client/src/bastionlab/torch/utils.py:316]: what is this change?

File: client/src/bastionlab/torch/remote_torch.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+@dataclass
+class RemoteDataset:
+    inputs: List[RemoteTensor]
+    labels: RemoteTensor
+    name: Optional[str] = "RemoteDataset-" + hashlib.sha256().hexdigest()[:5]
+    privacy_limit: Optional[float] = -1.0
+
+    @property
+    def trace_input(self):
```
Comment (cchudant) [inline client/src/bastionlab/torch/remote_torch.py:117]: should this be public api?

Comment (cchudant) [reply client/src/bastionlab/torch/remote_torch.py:117]: ah it was also in the old code..

File: protos/bastionlab.proto [hunk 5]
```diff
@@ -3,6 +3,22 @@ package bastionlab;
...
+    string description = 3;
+    bytes meta = 4;
+}
+
+message TensorMetaData {
+    // This message represents metadata about Tensor.
+    // input_dtype --> tch::Tensor::kind() which we map onto PyTorch dtypes
+    // input_shape --> List of integers representing the shape. This is transformed into PyTorch Shape
+    
+    repeated int64 input_shape = 1;
```
Comment (cchudant) [inline protos/bastionlab.proto:18]: why "input"? maybe just shape/dtype?

File: protos/bastionlab.proto [hunk 6]
```diff
@@ -3,6 +3,22 @@ package bastionlab;
 
 message Empty {}
 
+message Reference {
```
Comment (cchudant) [inline protos/bastionlab.proto:6]: cool! :+1:

File: protos/bastionlab_conversion.proto [hunk 7]
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
+
+package bastionlab_conversion;
+import "bastionlab.proto";
+
+message ToTensor {
```
Comment (cchudant) [inline protos/bastionlab_conversion.proto:6]: what is this?

Comment (cchudant) [reply protos/bastionlab_conversion.proto:6]: ah i understand

File: protos/bastionlab_polars.proto [hunk 8]
```diff
@@ -32,18 +32,27 @@ message FetchChunk {
...
 
-message Query {
-    string composite_plan = 1;
+message QueryBytes {
+    bytes data = 1;
 }
 
 message Empty {}
 
+message SplitRequest {
```
Comment (cchudant) [inline protos/bastionlab_polars.proto:41]: same comment as python `tain_test_split`

File: server/bastionlab_common/src/common_conversions.rs [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    let mut ldf = LazyFrame::default();
+    ldf.logical_plan = plan;
+    ldf
+}
+
+pub fn to_status_error<T, E: Error>(input: Result<T, E>) -> Result<T, Status> {
+    input.map_err(|err| Status::aborted(err.to_string()))
+}
+
+pub fn load_udf(udf: String) -> Result<CModule, Status> {
```
Comment (cchudant) [inline server/bastionlab_common/src/common_conversions.rs:185]: do we support udfs?

File: server/bastionlab_polars/src/composite_plan.rs [hunk 10]
```diff
@@ -18,7 +21,15 @@ pub struct CompositePlan(Vec<CompositePlanSegment>);
 #[derive(Debug, Serialize, Deserialize)]
 pub enum CompositePlanSegment {
     PolarsPlanSegment(LogicalPlan),
-    UdfPlanSegment { columns: Vec<String>, udf: String },
+    StringTransformerPlanSegment {
+        columns: Vec<String>,
+        model: String,
+        add_special_tokens: u8,
+    },
+    UdfPlanSegment {
```
Comment (cchudant) [inline server/bastionlab_polars/src/composite_plan.rs:29]: do we support that? (python & rust side)

File: server/bastionlab_polars/src/lib.rs [hunk 11]
```diff
@@ -531,4 +547,69 @@ impl PolarsService for BastionLabPolars {
         );
         Ok(Response::new(Empty {}))
     }
+    async fn split(
```
Comment (cchudant) [inline server/bastionlab_polars/src/lib.rs:550]: same train_test_split comment

File: tests/test_training.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        test_labels = test_labels.to_tensor()
+
+        in_features = train_inputs.shape[-1]
+        dtype = train_inputs.dtype
+        model = make_model(in_features, dtype)()
+
+        train_dataset = RemoteDataset(inputs=[train_inputs], labels=train_labels)
+        test_dataset = RemoteDataset(inputs=[test_inputs], labels=test_labels)
+
+        remote_learner = connection.client.torch.RemoteLearner(
```
Comment (cchudant) [inline tests/test_training.py:109]: cool!! thanks for making tests :)

File: client/src/bastionlab/polars/remote_polars.py [hunk 13]
```diff
@@ -1327,3 +1396,84 @@ def __map(self: LDF, func, **kwargs) -> None:
...
+    res = [
+        FetchableLazyFrame._from_reference(
+            _train_rdf._meta._polars_client, ref
+        ).to_array()
+        for ref in res.list
+    ]
+    return res
+
+
+class RemoteArray(RemoteLazyFrame):
```
Comment (cchudant) [inline client/src/bastionlab/polars/remote_polars.py:1458]: i still dont get what remote array is :sweat_smile: 

if i understand correctly, it's something that is client-only, for remotedataframe.to_array().to_tensor()? cant we just have remotedataframe.to_tensor?
i cant see any RemoteArray in the rust code

Comment (kbamponsem) [reply client/src/bastionlab/polars/remote_polars.py:1458]: You can ask Lucas about RemoteArray's.
He did the framing

File: server/bastionlab_polars/src/lib.rs (hunk 11 above)
Comment (kbamponsem) [reply server/bastionlab_polars/src/lib.rs:550]: Well, as I said, the features are inter-related and I cannot use RemoteTokenizers without the data-conversion feature.
And splitting cannot be implemented without RemoteArray's (data-conversion) feature.

File: server/bastionlab_polars/src/composite_plan.rs (hunk 10 above)
Comment (kbamponsem) [reply server/bastionlab_polars/src/composite_plan.rs:29]: Do we support what exactly?

File: server/bastionlab_common/src/common_conversions.rs (hunk 9 above)
Comment (kbamponsem) [reply server/bastionlab_common/src/common_conversions.rs:185]: Well, I think that's what `apply_udf` use. And if you want more info, ask Lucas.

File: protos/bastionlab_conversion.proto [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+message ToDataFrame {
+    repeated string inputs_col_names = 1;
+    string labels_col_name = 2;
+    string inputs_conv_fn = 3;
+    string labels_conv_fn = 4;
+    string identifier = 5;
+}
+
+service ConversionService {
+    rpc ConvToTensor(ToTensor) returns (bastionlab.Reference) {}
```
Comment (cchudant) [inline protos/bastionlab_conversion.proto:19]: (i am reviewing with lucas)
Now that we have bastionlab.Reference, I think we can move this back in the polars protobuf directly, what do you think? We won't have import problems anymore, do you think that would work?
Sorry if this means changing things a lot :sweat_smile:

File: client/src/bastionlab/polars/remote_polars.py (hunk 13 above)
Comment (cchudant) [reply client/src/bastionlab/polars/remote_polars.py:1458]: Okay thanks, i talked with Lucas

Comment (cchudant) [reply client/src/bastionlab/polars/remote_polars.py:1458]: Do remote lazy frames support .to_tensor directly ? (as sugar for .to_array().to_tensor())

**Review by dhalf — APPROVED** [review]

**Pull request merged**